src/category_theory/connected.lean

```diff
@@ -0,0 +1,374 @@
+component'.
+
+This allows us to show that the functor X ⨯ - preserves connected limits.
+-/
+class connected (J : Type v₂) [𝒥 : category.{v₁} J] extends inhabited J :=
+(iso_constant : Π {α : Type v₂} (F : J ⥤ discrete α), F ≅ (functor.const J).obj (F.obj default))
+end connected
+
+section J
+variables {J : Type v₂} [𝒥 : category.{v₁} J]
+include 𝒥
+
+/-- If J is connected, any functor to a discrete category is constant on the nose. -/
+lemma any_functor_eq_constant [conn : connected J] {α : Type v₂} (F : J ⥤ discrete α) :
+  F = (functor.const J).obj (F.obj (default J)) :=
+begin
+  apply functor.ext _ _,
+    intro X,
+    have z := conn.iso_constant,
+    exact ((z F).hom.app X).down.1,
+  intros, apply subsingleton.elim
+end
+
+/-- If any functor to a discrete category is constant on objects, J is connected. -/
+def connected.of_any_functor_const_on_obj [inhabited J]
+  (h : ∀ {α : Type v₂} (F : J ⥤ discrete α), ∀ (j : J), F.obj j = F.obj (default J)) :
+  connected J :=
+{ iso_constant := λ α F, nat_iso.of_components (λ B, eq_to_iso (h F B)) (λ _ _ _, subsingleton.elim _ _) }
+
+/--
+If `J` is connected, then given any function `F` such that the presence of a
+morphism `j₁ ⟶ j₂` implies `F j₁ = F j₂`, we have that `F` is constant.
+This can be thought of as a local-to-global property.
+
+The converse is shown in `connected.of_constant_of_preserves_morphisms`
+-/
+lemma constant_function_of_preserves_morphisms [connected J] {α : Type v₂} (F : J → α) (h : ∀ (j₁ j₂ : J) (f : j₁ ⟶ j₂), F j₁ = F j₂) (j : J) :
+  F j = F (default J) :=
+@congr_arg (J ⥤ discrete α) _ _ _ (λ t, t.obj j) (any_functor_eq_constant { obj := F, map := λ _ _ f, eq_to_hom (h _ _ f) })
+
+/--
+`J` is connected if: given any function `F : J → α` which is constant for any
+`j₁, j₂` for which there is a morphism `j₁ ⟶ j₂`, then `F` is constant.
+This can be thought of as a local-to-global property.
+
+The converse of `constant_function_of_preserves_morphisms`.
+-/
+def connected.of_constant_of_preserves_morphisms [inhabited J]
+  (h : ∀ {α : Type v₂} (F : J → α), (∀ {j₁ j₂ : J} (f : j₁ ⟶ j₂), F j₁ = F j₂) → (∀ j : J, F j = F (default J))) :
+  connected J :=
+connected.of_any_functor_const_on_obj (λ _ F, h F.obj (λ _ _ f, (F.map f).down.1))
+
+/--
+An inductive-like property for the objects of a connected category.
+If `default J` is in the set `p`, and `p` is closed under morphisms of `J`,
+then `p` contains all of `J`.
+
+The converse is given in `connected.of_induct`.
+-/
+lemma induct_on_objects [connected J] (p : set J) (h0 : default J ∈ p)
+  (h1 : ∀ {j₁ j₂ : J} (f : j₁ ⟶ j₂), j₁ ∈ p ↔ j₂ ∈ p) (j : J) :
+  j ∈ p :=
+begin
+  injection (constant_function_of_preserves_morphisms (λ k, ulift.up (k ∈ p)) (λ j₁ j₂ f, _) j) with i,
+  rwa i,
+  dsimp,
+  exact congr_arg ulift.up (propext (h1 f)),
+end
+
+/--
+In other words, this says that any maximal connected component of J containing the default must be all of J.
```

Comment:
Best if doc-strings are readable in isolation (e.g. as pop-up tooltips), so probably shouldn't start with "In other words"